rotkehlchen/chain/evm/decoding/odos/v2/decoder.py

```diff
@@ -0,0 +1,228 @@
+            except DeserializationError as e:
+                log.error(
+                    f'Failed to deserialize Odos event {context.tx_log=} at '
+                    f'{context.transaction.tx_hash.hex()} due to {e}',
+                )
+                return DEFAULT_DECODING_OUTPUT
+        else:
+            return DEFAULT_DECODING_OUTPUT
+
+        if not self.base.is_tracked(decoded_data[0]):
+            return DEFAULT_DECODING_OUTPUT
+
+        input_tokens = self._resolve_tokens_data(token_amounts=decoded_data[1], token_addresses=decoded_data[2])  # noqa: E501
+        output_tokens = self._resolve_tokens_data(token_amounts=decoded_data[3], token_addresses=decoded_data[4])  # noqa: E501
+
+        in_events, out_events, fee_events = [], [], []
+        for event in context.decoded_events:
+            if (
+                ((input_amount := input_tokens.get(event.asset.identifier)) == event.balance.amount) or  # noqa: E501
+                input_amount == ZERO  # this means https://docs.odos.xyz/product/sor/v2/#max-balance-swapping  # noqa: E501
+            ) and (
+                (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ) or (
+                    event.event_type == HistoryEventType.TRADE and
+                    event.event_subtype == HistoryEventSubType.SPEND
+                )
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.balance.amount} {event.asset.symbol_or_name()} in Odos v1'  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.address = self.router_address
+                event.counterparty = CPT_ODOS_V2
+                out_events.append(event)
+
+            elif output_tokens.get(event.asset.identifier) == event.balance.amount and (
+                (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ) or (
+                    event.event_type == HistoryEventType.TRADE and
+                    event.event_subtype == HistoryEventSubType.RECEIVE
+                )
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.notes = f'Receive {event.balance.amount} {event.asset.symbol_or_name()} as the result of a swap in Odos v1'  # noqa: E501
+                event.address = self.router_address
+                event.counterparty = CPT_ODOS_V2
+                in_events.append(event)
+
+        # fees are the amount of out tokens that router received - the amount that router sent
+        router_holding_amount: defaultdict[Asset, FVal] = defaultdict(FVal)
+        for tx_log in context.all_logs:
```

Comment:
maybe you can abstract this part in its own method and call it `_calculate_router_fee`?